### PR TITLE
Fix for zero average order value on charts

### LINKED ISSUE #15262: Average order value graph flattens out to value 0
Hello,

we have run into the same problem as mentioned in https://github.com/matomo-org/matomo/issues/7851

When looking at Ecommerce/Overview the average order value graph sticks to zero regardless that the summary and API call report a value other than zero.

ecommerce_evolution
ecommerce_summary

Our Matomo (3.13) setup is still the same as described in https://github.com/matomo-org/matomo/issues/15240

Please let me know, if you need any further information and thanks for your efforts in advance!

### PR DESCRIPTION
**Pull request by bx80**
### Description:

Fixes #15262 

The average order value on the evolution overview charts was always shown as zero, this was caused by pre-formatted values in the datatable breaking the jqplot data generation.

This fix attempts to solve this by:
- Respecting evolution chart visualization calls which set `format_metrics = 0` (only setting a `format_metrics` default value if there are no request override parameters)
- Having evolution charts set `format_metrics = 0`
- Passing the series units to the jqplot data generation `setAxisYValues` method
- If the series unit is a percent then convert chart data values to display percentages.
- jqplot chart number formatting currently doesn't fully support internationalization (eg. forcing a dot as a decimal point separator), all series formatting units were being added as a postfix, I added a minor tweak to show dollar and pound currency symbols as prefixes. This is obviously not ideal but should be a slight improvement for some people, if we had a reliable way to determine the currency symbol position then we could possibly add that as metadata to `core/Metrics::getUnit()` and make it available to the jqplot formatting code.

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

File: plugins/CoreVisualizations/JqplotDataGenerator/Chart.php [hunk 1]
```diff
@@ -90,7 +101,11 @@ public function setAxisYUnits($yUnits)
 
         // generate jqplot axes config
         foreach ($axesIds as $unit => $axisId) {
-            $this->axes[$axisId]['tickOptions']['formatString'] = '%s' . $unit;
+            if ($unit === '$' || $unit === '£') {
+                $this->axes[$axisId]['tickOptions']['formatString'] = $unit . '%s';
+            } else {
+                $this->axes[$axisId]['tickOptions']['formatString'] = '%s' . $unit;
+            }
         }
```
Comment (sgiehl) [inline plugins/CoreVisualizations/JqplotDataGenerator/Chart.php]: Nice approach, but that won't be correct for all currencies and languages. I would try to do something like this maybe:

```php
        $currencies = StaticContainer::get('Piwik\Intl\Data\Provider\CurrencyDataProvider')->getCurrencyList();
        $currencies = array_column($currencies, 0);

        // generate jqplot axes config
        foreach ($axesIds as $unit => $axisId) {
            if ($unit === '%') {
                $this->axes[$axisId]['tickOptions']['formatString'] = str_replace('0', '%s', NumberFormatter::getInstance()->formatPercent(0, 0, 0));
            } else if (in_array($unit, $currencies)) {
                $this->axes[$axisId]['tickOptions']['formatString'] = str_replace('0', '%s', NumberFormatter::getInstance()->formatCurrency(0, $unit, 0));
            } else {
                $this->axes[$axisId]['tickOptions']['formatString'] = '%s' . $unit;
            }
        }
```

File: core/API/Request.php [hunk 2]
```diff
@@ -105,7 +105,8 @@ public static function getRequestArrayFromString($request, $defaultRequest = nul
                 $defaultRequest['segment'] = $requestRaw['segment'];
             }
 
-            if (!isset($defaultRequest['format_metrics'])) {
+            // Only default to formatting metrics if the request doesn't contain the format metrics parameter
+            if (!isset($defaultRequest['format_metrics']) && !isset($request['format_metrics'])) {
```
Comment (sgiehl) [inline core/API/Request.php]: Does that really change anything? If `format_metrics` is set in `$request` it should overwrite the value in `$defaultRequest` in the resulting array a couple of line below: `$requestArray = $requestParsed + $defaultRequest;`

File: plugins/Goals/Controller.php [hunk 3]
```diff
@@ -251,6 +252,10 @@ public function getEvolutionGraph(array $columns = array(), $idGoal = false, arr
         $langString = $idGoal ? 'Goals_SingleGoalOverviewDocumentation' : 'Goals_GoalsOverviewDocumentation';
         $view->config->documentation = $this->translator->translate($langString, '<br />');
 
+        if ($view instanceof Evolution) {
+            $view->requestConfig->request_parameters_to_modify['format_metrics'] = 0;
+        }
```
Comment (sgiehl) [inline plugins/Goals/Controller.php:257]: Actually I was wondering why only the Average Order Value is broken, but the problem seems that that metric is added as processed metric, while all other metrics are "normal" metrics, which causes the formatter to format only that one.
Guess that's yet another issue of not having a clean metric system, where all metrics in a datatable have a metric class, that might format if needed or not.
Might be something we could try to improve for Matomo 5...

**Review by sgiehl — CHANGES_REQUESTED** [review]
Seems to work as expected overall. Left some comments for further improvements.

File: plugins/CoreVisualizations/JqplotDataGenerator/Chart.php (hunk 1 above)
Comment (bx80) [reply plugins/CoreVisualizations/JqplotDataGenerator/Chart.php]: That's a lot better, I tested it with languages that place the currency symbol first and last, regardless of the actual symbol it was positioned in the correct place :+1:

File: core/API/Request.php (hunk 2 above)
Comment (bx80) [reply core/API/Request.php]: You're correct, that's not necessary, removed!

File: plugins/Goals/Controller.php (hunk 3 above)
Comment (bx80) [reply plugins/Goals/Controller.php:257]: Definitely a good candidate for Matomo 5. There seem to be a lot of issues at the moment caused by formatted/unformatted metrics, we've fixed quite a few but the whole system would benefit from some consistency :+1:

File: core/API/Request.php (hunk 2 above)
Comment (bx80) [reply core/API/Request.php]: I've added this back as without it the `SingleMetricView` dashboard widget loses it's formatting and those UI tests fail, it seems to be forcing `format_metrics = 'bc'` even if `format_metrics = 'all'` has been passed.

Comment (sgiehl) [reply core/API/Request.php]: I guess the check if it exists in `$request` would be unneeded. But it actually doesn't hurt as well.
I guess the problem here is (again) our magic api method `API.get`, which is used in the `SingleMetricView`. The `format_metrics` param doesn't seem to be in `$request`, but in `$defaultRequest`, where it lands as it is set in `$_GET`. 
Let's just keep that check and hope we can improve all the number formatting in Matomo 5

**Review by sgiehl — APPROVED** [review]

**Pull request merged**